File: src/main/java/me/alegian/thavma/impl/client/gui/thaumonomicon/grid/GridRenderable.java
package me.alegian.thavma.impl.client.gui.thaumonomicon.grid;

import me.alegian.thavma.impl.client.T7GuiGraphics;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.resources.ResourceLocation;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.api.distmarker.OnlyIn;

// represents any element drawn in the Grid, even if it doesn't directly snap to it
@OnlyIn(Dist.CLIENT)
public class GridRenderable {
  private final ResourceLocation texture;
  private final int x;
  private final int y;
  private final int sizeX;
  private final int sizeY;
  private final int rotationDegrees;

  public GridRenderable(ResourceLocation texture, int x, int y, int sizeX, int sizeY, boolean flip, int rotationDegrees) {
    this.texture = texture;
    this.x = x;
    this.y = y;
    this.sizeX = flip ? -sizeX : sizeX;
    this.sizeY = flip ? -sizeY : sizeY;
    this.rotationDegrees = rotationDegrees;
  }

  public GridRenderable(ResourceLocation texture, int x, int y, int sizeX, int sizeY) {
    this(texture, x, y, sizeX, sizeY, false, 0);
  }

  public GridRenderable(ResourceLocation texture, int x, int y, int rotationDegrees) {
    this(texture, x, y, 1, 1, false, rotationDegrees);
  }

  public GridRenderable(ResourceLocation texture, int x, int y, boolean flip, int rotationDegrees) {
    this(texture, x, y, 1, 1, flip, rotationDegrees);
  }

  public GridRenderable(ResourceLocation texture, int x, int y) {
    this(texture, x, y, 1, 1);
  }

  public void render(GuiGraphics guiGraphics, int cellSize, double scrollX, double scrollY, boolean hovered, float tickDelta) {
    final var graphics = new T7GuiGraphics(guiGraphics);
    double xPos = cellSize * (x - sizeX / 2f);
    double yPos = cellSize * (y - sizeY / 2f);

    graphics.push();
    graphics.translateXY((float) -scrollX, (float) -scrollY);
    graphics.rotateZ(rotationDegrees);
    graphics.translateXY((float) xPos, (float) yPos);
    graphics.blitSimple(
        texture,
        0,
        0,
        cellSize * sizeX,
        cellSize * sizeY
    );
    graphics.pop();
  }
}

File: src/main/java/me/alegian/thavma/impl/client/gui/thaumonomicon/grid/GridRenderable.kt
package me.alegian.thavma.impl.client.gui.thaumonomicon.grid

import me.alegian.thavma.impl.client.util.rotateZ
import me.alegian.thavma.impl.client.util.translateXY
import me.alegian.thavma.impl.client.util.usePose
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.resources.ResourceLocation

// represents any element drawn in the Grid, even if it doesn't directly snap to it
// clockwise: 
open class GridRenderable(private val texture: ResourceLocation, private val x: Int, private val y: Int, sizeX: Int = 1, sizeY: Int = 1, flip: Boolean = false, private val rotationDegrees: Int = 0) {
  private val sizeX = if (flip) -sizeX else sizeX
  private val sizeY = if (flip) -sizeY else sizeY

  constructor(texture: ResourceLocation, x: Int, y: Int, rotationDegrees: Int) : this(texture, x, y, 1, 1, false, rotationDegrees)

  constructor(texture: ResourceLocation, x: Int, y: Int, flip: Boolean, rotationDegrees: Int) : this(texture, x, y, 1, 1, flip, rotationDegrees)

  fun render(graphics: GuiGraphics, cellSize: Int, hovered: Boolean, tickDelta: Float) {
    val xPos = (cellSize * (x - sizeX / 2f)).toDouble()
    val yPos = (cellSize * (y - sizeY / 2f)).toDouble()

    graphics.usePose {
      rotateZ(rotationDegrees.toFloat())
      translateXY(xPos, yPos)
      graphics.blit(
        texture,
        0,
        0,
        0,
        0f,
        0f,
        cellSize * sizeX,
        cellSize * sizeY,
        cellSize * sizeX,
        cellSize * sizeY
      )
    }
  }
}
